camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/SelfManagedAuthentication.java

```diff
@@ -69,17 +69,21 @@ public Map.Entry<String, String> getTokenHeader(Product product) {
     if (tokens.containsKey(product)) {
       token = tokens.get(product);
     } else {
-      Identity identity = identityConfig.get(product).getIdentity();
-      String audience = jwtConfig.getProduct(product).getAudience();
-      Tokens identityTokens = identity.authentication().requestToken(audience);
-      try {
-        identity.authentication().verifyToken(identityTokens.getAccessToken());
-      } catch (TokenExpiredException exception) {
-        identityTokens = identity.authentication().renewToken(identityTokens.getRefreshToken());
-      }
-      tokens.put(product, identityTokens.getAccessToken());
-      token = tokens.get(product);
+      token = getIdentityToken(product);
     }
     return new AbstractMap.SimpleEntry<>("Authorization", "Bearer " + token);
   }
+
+  private String getIdentityToken(Product product) {
+    Identity identity = identityConfig.get(product).getIdentity();
+    String audience = jwtConfig.getProduct(product).getAudience();
+    Tokens identityTokens = identity.authentication().requestToken(audience);
+    try {
+      identity.authentication().verifyToken(identityTokens.getAccessToken());
+    } catch (TokenExpiredException exception) {
+      identityTokens = identity.authentication().renewToken(identityTokens.getRefreshToken());
+    }
```

Comment:
I still dont get why we renew token we just created here?  Or does the `identity` cache it internally?

I would recommend to take a look at the Identity SDK internals: https://github.com/camunda-cloud/identity/blob/db2dae3da92bbbfc34e646297e4bdd6ba04f07ec/sdk-java/src/main/java/io/camunda/identity/sdk/authentication/AbstractAuthentication.java#L80

Reply:
So the issue this is trying to solve is because of the following problem:

Identity code internals[ doesn't provide a way ](https://github.com/camunda-cloud/identity/issues/2496#issuecomment-1904273555)for its users to remove expired cache token internally when you are using the Identity SDK. I requested Identity team to expose a method that will either delete the token or refresh it. 

If we don't have the verify / renew token code here (L81-L85), what would happen is we would keep getting expired tokens since `identityTokens.getAccessToken()` always pull it from their cache and its not refreshing it automatically.

> I still dont get why we renew token we just created here?

We renew it since Identity internals doesn't renew their tokens when first created. So even if the method here says `getIdentityToken(Product product)`, it always pull from cache. It is our responsibility to check if it was expired and if its then refresh it.

Once Identity provided a method to get a fresh and update their internal cache, then the entire `try-catch` block can be remove, and thus avoids us checking if token is expired


Does this make sense?

Reply:
Ok, thanks for the explanation. I thought we might rather need to do it when a client receives an error from the server. Not sure why the SDK would return an invalid token from the cache.

Could you create a ticket in the Identity SDK for this issue: https://github.com/camunda-cloud/identity/issues so the SDK can provide this functionality?
cc @Ben-Sheppard

Reply:
Raised the [issue](https://github.com/camunda-cloud/identity/issues/2496) on Identity

Reply:
I would also need to update some of the code here since I got to know that M2M calls will never have refreshTokens so this line has to go away

`identityTokens = identity.authentication().renewToken(identityTokens.getRefreshToken())`

Will refactor and do some testing for expiration scenarios for this